linkerd/proxy/tap/src/registry.rs

```diff
@@ -0,0 +1,58 @@
+use crate::iface;
+use std::{
+    sync::{Arc, Mutex},
+    time::Duration,
+};
+use tokio::{sync::watch, time};
+use tracing::trace;
+
+#[derive(Clone, Debug)]
+pub struct Registry<T> {
+    inner: Arc<Mutex<Inner<T>>>,
+    taps_recv: watch::Receiver<Vec<T>>,
+}
+
+#[derive(Debug)]
+struct Inner<T> {
+    taps: Vec<T>,
+    taps_send: watch::Sender<Vec<T>>,
+}
+
+impl<T> Registry<T>
+where
+    T: iface::Tap + Clone,
+{
+    pub fn new() -> Self {
+        let (taps_send, taps_recv) = watch::channel(vec![]);
+        let inner = Inner {
+            taps: Vec::default(),
+            taps_send,
+        };
+        Self {
+            inner: Arc::new(Mutex::new(inner)),
+            taps_recv,
+        }
+    }
+
+    pub fn get_taps(&self) -> Vec<T> {
+        self.taps_recv.borrow().clone()
+    }
+
+    pub fn register(&self, tap: T) {
+        if let Ok(mut inner) = self.inner.lock() {
+            inner.taps.push(tap);
+            let _ = inner.taps_send.broadcast(inner.taps.clone());
+        }
+    }
+
+    pub async fn clean(self) {
+        loop {
+            if let Ok(mut inner) = self.inner.lock() {
+                let count = inner.taps.len();
+                inner.taps.retain(|tap| tap.can_tap_more());
+                trace!("retained {} of {} taps", inner.taps.len(), count);
+            }
+            time::delay_for(Duration::from_millis(100)).await;
+        }
+    }
```

Comment:
* Let's pass in an `interval: Duration` rather than hardcode it here.
* This task never completes. We'll need to select on `tx.closed()` -- I think this means that the tx should be stored on the registry and not within the mutex...
* The task should probably start by waiting, rather than cleaning immediately, but that's fairly minor.

Reply:
Better yet, you could pass in a `wakeup: impl Stream<Output = ()>` to completely decouple the decision of when to clean from the task.

Reply:
Nevermind re: shutting this task down.. because the registry holds the receiver we can't know when all instances have been dropped.  If we'd like to support graceful cleanup (which isn't at all required, functionally), we'd have to decouple `clean` from `Registry`...